### PR TITLE
Integrate quantities

### PR DESCRIPTION
**Pull request by sonVishal**
By setting the `integrateQuants` flag to `yes` while compiling we can enable the computation and storage of 1st order time integration of the variables.

To enable or disable a variable for computation and storage a new section has been incorporated in the parameter file. The following example illustrates the new section:
```
&Postprocessing
IntegrationMask = 0 1 0 1 1 0 0 1 1  ! Integrate and output these vars
/
```
In this example, the variables `sigma_xx`, `sigma_xz`, `sigma_zz` and `u` will not be integrated and stored.

### CODE REVIEW

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 1]
```diff
@@ -285,7 +296,16 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
+		int numLowVars = 0;
+		if (pstrain && !integrals) {
+			numLowVars = WaveFieldWriterExecutor::NUM_LOWVARIABLES;
+		} else if (integrals && !pstrain) {
+			numLowVars = m_numIntegratedVariables;
+		} else {
+			numLowVars = WaveFieldWriterExecutor::NUM_LOWVARIABLES + m_numIntegratedVariables;
+		}
+
+		for (unsigned int i = 1; i < numLowVars; i++)
```
Comment (rettenbs) [inline src/ResultWriter/WaveFieldWriter.cpp]: You can write this with:
`numLowVars = 0`
`if (pstrain) numLowVars += WaveFieldWriterExecutor::NUM_LOWVARIABLES;`
`if (integrals) numLowVars += m_numIntegratedVariables;`

File: src/ResultWriter/WaveFieldWriter.h [hunk 2]
```diff
@@ -248,6 +258,33 @@ class WaveFieldWriter : private async::Module<WaveFieldWriterExecutor, WaveField
 			}
 		}
 
+		if (m_integrals) {
+			unsigned int offset = 0;
+			unsigned int flagOffset = WaveFieldWriterExecutor::NUM_LOWVARIABLES;
+		if (m_pstrain) {
+				offset = WaveFieldWriterExecutor::NUM_LOWVARIABLES;
+				flagOffset = 0;
+			}
```
Comment (rettenbs) [inline src/ResultWriter/WaveFieldWriter.h]: This is a bit difficult to understand. It is not clear what offset, flagOffset, nextId and i are for. You do not need `flagOffset` (see next comment) and `offset` can be set between line 244 and 259. Saves one additional if statement.

File: src/ResultWriter/WaveFieldWriter.h [hunk 3]
```diff
@@ -248,6 +258,33 @@ class WaveFieldWriter : private async::Module<WaveFieldWriterExecutor, WaveField
...
+
+			nextId = offset;
+			for (unsigned int i = offset; i < offset+WaveFieldWriterExecutor::NUM_INTEGRATED_VARIABLES; i++) {
```
Comment (rettenbs) [inline src/ResultWriter/WaveFieldWriter.h]: Iterate form 0 to `WaveFieldWriterExecutor::NUM_INTEGRATED_VARIABLES`. You can remove `flagOffset` completely and always use `WaveFieldWriterExecutor::NUM_LOWVARIABLES`.

File: src/ResultWriter/WaveFieldWriterExecutor.h [hunk 4]
```diff
@@ -280,6 +307,8 @@ class WaveFieldWriterExecutor
 
 public:
 	static const unsigned int NUM_LOWVARIABLES = 7;
+	static const unsigned int NUM_INTEGRATED_VARIABLES = 9;
+	static const unsigned int NUM_TOTALLOWVARS = 16;
```
Comment (rettenbs) [inline src/ResultWriter/WaveFieldWriterExecutor.h]: `static const unsigned int NUM_TOTALLOWVARS = NUM_LOWVARIABLES + NUM_INTEGRATED_VARIABLES`

In addition, the names `NUM_TOTALLOWVARS` and `NUM_LOWVARIABLES` are a bit misleading. `NUM_LOWVARIABLES` should be the total number of variables and `NUM_PLASTICITY_VARIABLES` should be the name for what was defined as `NUM_LOWVARIABLES` before.

File: src/ResultWriter/PostProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#include "PostProcessor.h"
```
Comment (rettenbs) [inline src/ResultWriter/PostProcessor.cpp]: Add the common header to all new files. You can copy/paste it from one of the existing files and replace the author with your name.

**Comment by sonVishal** [comment]
I have made the relevant changes and taken care that GCC doesn't complain. Tested the changes with TPV33 and TPV27 benchmarks with and without the integration.

Let me know if any more changes are required.

File: src/ResultWriter/WaveFieldWriter.h [hunk 6]
```diff
@@ -278,7 +272,7 @@ class WaveFieldWriter : private async::Module<WaveFieldWriterExecutor, WaveField
 #endif // _OPENMP
 				for (unsigned int j = 0; j < m_numLowCells; j++)
 					managedBuffer[j] = m_integrals[m_map[j]
-							* m_numIntegratedVariables + nextId-offset];
+							* m_numIntegratedVariables + nextId];
```
Comment (rettenbs) [inline src/ResultWriter/WaveFieldWriter.h]: I have not tested it, but does this really work when pstrain is enabled?

**Pull request merged**